### ISSUE TITLE
Serialized CID sessions could become invalid when new mbedtls version is taken into use

### ISSUE THREAD

**Issue by samhaa01** [ARMmbed/mbedtls#3466]
### Description
- Type: Feature Request
- Priority: Major

---------------------------------------------------------------
## Bug

**OS**  
linux / MacOS

**mbed TLS build:**  
Version: mbedtls-2.22.0  
OS version: openjdk:8u232-jre-slim (docker)  
Configuration: please attach config.h file where possible  
Compiler and options (if you used a pre-built binary, please indicate how you obtained it):  
Additional environment information:  

**Peer device TLS stack and version**  
mbed TLS  
Version:  
mbedtls-2.22.0

**Expected behavior**   

At the moment we have a known limitation with mbedTLS CID implementation and it is that the session serialization is not compatible between session. So, whenever we take a new version of the mbedTLS library into use, the server is not able to decrypt the traffic coming from devices which is using an existing CID created with the previous version.
 
**Actual behavior**  
Client using CID generated with old mbedtls version is not able to connect.

**Steps to reproduce**  
Connect to server using old mbedtls with empty CID. Client will receive a CID from server, which it can use to connect later.
Disconnect or have the session to be disconnected using time-out. This will make the server save the CID session.
Update server and take new mbedtls into use.
Try to connect using the client to the new server. --> connecting fails. To recover the client needs to handshake again. Purpose for CID is to avoid handshaking when connecting again. 
----------------------------------------------------------------
## Enhancement\Feature Request

**Justification - why does the library need this feature?**  
If server updates mbedtls version, the clients will go offline until they re-handshake.

**Suggested enhancement**  

-----------------------------------------------------------------

**Comment by mpg** [comment]
Hi,

while I don't disagree about the requested enhancement, I'd like to try and clarify some concepts and terminology.

- CID, or Connection Identifiers, is an extension to the DTLS protocol that is [standardized by the IETF](https://tools.ietf.org/html/draft-ietf-tls-dtls-connection-id-07). As such, it is interoperable between different versions of Mbed TLS that implement this extension, and even between Mbed TLS and any other implementation that implements this extension. (With the minor caveat that currently it's still a draft, not an RFC yet, so there could be changes between the current version an the final version of the specification, though that seems unlikely at this point.)

- Context serialisation, that is serialisation of an `ssl_context` structure (which is different from serialisation of an `ssl_session` structure, so I think it's really preferable to say "context serialisation" rather than "session serialisation" here), on the other hand is an out-of-protocol feature specific to Mbed TLS. When one of the parties in a connection uses this feature, it's completely transparent to the peer, it really is just a local optimisation (much like the OS swapping pages from RAM to disk is transparent to user-space). This is the part which comes with limitations about upgrading: the same build (version + config.h options) of Mbed TLS must be used for serialisation and deserialisation (with the same `ssl_config` too).

Those two features are logically quite independent from each other. I understand that you use them both in conjunction to achieve very long-lived DTLS associations, so from your perspective it makes sense to call that "CID sessions", but really from a TLS perspective they're independent things (and the word "session" already has a specific meaning in the TLS standard, that is yet another thing). So I'm taking the liberty of editing the issue title to remove references to CID, and refer to "serialized contexts" instead.

**Comment by mpg** [comment]
In the meantime, I'm wondering if you could apply a gradual upgrade strategy for your server instances. For example, when you're upgrading from version N to version N+1 of mbed TLS, serve all new clients with version N+1 instances, but keep a few version N instances around to serve clients with existing connections, until their connection expires and they have to re-handshake anyway.

I don't know how long-lived your connections are, but Mbed TLS releases are not too frequent, so perhaps this is workable. Even if your connections are usually too long-lived to keep version N instances around for so long, and you have to force re-handshakes to get rid of the version N instances, at least this could help spread the cost of those re-handshakes over a few days instead of having all of them at once when you upgrade your server software.

Please note I'm not disputing the usefulness of the enhancement you're suggesting, I'm just trying to explore options for coping in the meantime.

**Comment by smolboarm** [comment]
Hi, Manuel.

The biggest problem that we can't "move" clients with serialized session to particular instances and keep them with different version. Because in front of our service there is load-balancer and we never know which pod will be hit by particular client.
I understand, that most of the problem is that ssl_context is serialized with version number and on deserialization there is a strict check for version match.
Also all the instances, in our case at the moment, updated at same time (tens of minutes).
I understand, that problem for deserialization is in fields offset/size matching, which _can_ depend on mbedtls version, when some fields added/changed/removed. And it definitely depends on configuration during compilation (which features / fields are actually included in compiled library).
Probably it worth to use some forma, where is no such hard dependencies, and where non-critical fields can be ommited from previous version. I.e. not offset base, but name based fields. Probably something like protobuf format (or field indexes, instead pure offsets).
Other way, which is much simplier from my point of view, is to keep deserialization for previous version. I.e. having two methods - deserialize_context, and deserialize_context_prev_ver, depending on actual version present in the data. That will solve our problems, but implies manual work on your side, for copying previous version deserialization code to such method every time new release is made. :(
So assuming we have same configuration when upgrading mbedtls version - it would be cool to have backward compatibility for previous version. But I'm not sure how to make it robust against configuration change. Here only protobuf/json like formats can help with filed's names or single time used indexes (single time per field introduction, so if field is removed - this index is never used again for new fields).

**Comment by mpg** [comment]
Hi Oleg,

thanks for considering my suggestion and explaining why it doesn't work for your use case.

We considered using named fields, but since the feature can also be useful on constrained devices (that may want to serialize a context to flash before going to deep sleep), an offset based format seemed more suitable there as it tends to result in more compact code and smaller serialized data.

> Other way, which is much simplier from my point of view, is to keep deserialization for previous version. I.e. having two methods - deserialize_context, and deserialize_context_prev_ver, depending on actual version present in the data.

Yes, that's the kind of thing we had in mind. Or have a single top-level function that branches to the appropriate function depending on the version, to be more user-friendly, but either way, some code on our side that decides what to do with missing or extra fields. Since some fields have security implications, it's likely that regardless of the format, there will be manual work on our side to check what to do - even with a name-based format if there's an missing name in the serialised format we need to decide on a safe default, if there's an extra name we need to decide if it's safe to ignore it - and in some cases there might be no safe fallback and we might have to just return an error. (Though those cases should be fairly rare and I don't think there's such a field right now, but I don't think that's a possibility we can completely exclude in the future either.)

I have another question for you: would you be interested in incremental progress on this? For example deserializing a context from the previous version would work with _some_ version upgrades but perhaps not _all_. I'm asking because there's a low-hanging fruit here: most of our releases actually don't change the format at all. So it would pretty easy to just replace the check for `serialized_version == current_version` by a check `serialized_version >= last_version_the_format_changed`. (For example, a quick git blame suggest we didn't change the format yet since we introduced this serialization feature.) The only additional work on use is to update that when we change the format, but since a superset of the format is already checked by our ABI checker script, that's actually not so much extra work. Then the result for you would be that _most_ version upgrades could re-use the old serialized data, while _some_ would still require a re-handshake - that is, until we implement a more complete solution.

Please note these are just my personal thoughts and if you're interested, the feasibility of that approach would have to be confirmed by other members of the team, but I first wanted to check if a partial solution - while waiting for a more complete one - would be of interest to you.

**Comment by smolboarm** [comment]
Hi, Manuel.

Yes, this approach was kind of obvious, that why I did not mentioned it explicitly (probably should have stated that explicitly). That's why word "can" was highlighted by me in the sentence: "I understand, that problem for deserialization is in fields offset/size matching, which can depend on mbedtls version, when some fields are added/changed/removed."

So yes, I, personally, vote for this approach with **format** version check instead of **library** version check. Also noticed that format does not actually change between version (at least for last few versions).

At least we will not need to force all clients to re-handshake on every library update.

**Comment by mpg** [comment]
Hi Oleg,

>  this approach with _format_ version check instead of _library_ version check

That's an excellent point, now that you mention it I'm not even sure why we included a check for the library version in the first place, instead of using a version number for the serialisation format. Perhaps just because it seemed convenient.

@hanno-arm Since you were involved in the design discussions about that, I'm curious to know your opinion if you happen to have some time to look at this.

**Comment by smolboarm** [comment]
Hi, Manuel.
I have one more probable idea, to mitigate configuration difference check even version is same. Format version can be composed from format version and configuration hash (only configuration which affects serialization format). Probably it can be done somehow (python? cmake?). But this is just an idea.

**Comment by jaudiger** [comment]
Hi all,

I encountered this issue recently after having updated mbed TLS to the latest version. Since no breaking change was existing between both version, I was wondering why this check was set up in the first place.

Checking is valid, I have nothing against that, I was just wondering why it was linked to the mbed TLS version. And if there is a possibility to decorrelate the increment of the mbed TLS version here.

**Comment by mpg** [comment]
Hi Jérémy,

Yes, I _think_ it would be possible to use "serialization format version" instead of using the Mbed TLS version here, which would only be incremented when the format actually changes, as opposed to with any new release (as previously suggested [above](https://github.com/Mbed-TLS/mbedtls/issues/3466#issuecomment-653428492)).

The obvious problem with that approach is the risk that we forget incrementing the format version when we change the format. This risk can probably be mitigated by an appropriate set of tests, which would have to be written.

I guess the reason we went for using the Mbed TLS version here was that we chose the simplest thing that was guaranteed to be correct, and thought we could improve incrementally starting from there. Evidently that did not quite happen as we moved to other things.

I'll raise this in the next team meeting and see what others think. In case we agree on a general strategy but lack the time to implement it, would any of you be interested in raising a PR?

**Comment by jaudiger** [comment]
Depending on the proposal and the impact in the mbed TLS code, it could be possible to create a PR on our side. But we'll see. At least thanks to raising it internally, this feature is a killer feature for us on our Server side. It adds the ability to serialize the security session of our DTLS peers. Even if there are some limitations currently: TLS serialization, handshake serialization and this issue. It matches some of our use cases.

We already maintained internally a patch to serialize TLS sessions, so far so good until now. The only downside to this patch is we have to maintain on the Application side some contexts to handle the streamable part of the TCP since we sometimes do not receive the complete buffer in one shoot. And that's why we didn't up streamed the patch because it could be better to handle it on mbed TLS side IMO.

The only remaining limitation for us (except this issue) is the no ability to serialize a security session that didn't finish its handshake. I never open an issue for this one, if you think this use case could make sense, I could fill one, let me know.

**Comment by mpg** [comment]
Ok, we didn't really have time to discuss it today as the meeting was pretty busy, with our upcoming relase, but I briefly mentioned it and hopefully we can discuss it more next time.

> This risk can probably be mitigated by an appropriate set of tests, which would have to be written.

Actually, before that we probably also need to think hard about the test strategy. I think coming up with a good test strategy is probably the hardest part in this.

>  this feature is a killer feature for us on our Server side

Glad to know this is useful :) And thanks for providing a summary of the limitations and how there are impacting you.

Regarding support for TLS (as opposed to DTLS), I'm afraid I didn't quite get the reason why you didn't try upstreaming your patches, since you say it would be better to handle things on the Mbed TLS side... I probably misunderstood something, could you clarify?

Regarding support for serialization before the handshake is complete, it's news to me that this is something people want. I'm a bit surprised, because as I understand it, serialization is mostly useful for long-lived connections to avoid having to re-handshake (as even a resumed handshake is expensive) but it would naively think that (1) the handshake itself would not be spread over a long period of time (like, at most a few minutes for slow devices) and (2) if we didn't complete the handshake yet, then there's less to be saved by resuscitating the SSL context. Apparently my reasoning does not match your experience, so I'm curious to know more.

(Perhaps you can open two new issues, one for TLS support and one for mid-handshake support, to keep this one focused on version upgrade. Disclaimer: by inviting you to create issues I'm not promising we'll work on them, but at least it provides a place to discuss, and if other people are interested they can chime in or +1 the issue so that we get a rough idea how many people are interested in each issue.)

**Comment by jaudiger** [comment]
Sorry for my late response.

> Regarding support for TLS (as opposed to DTLS), I'm afraid I didn't quite get the reason why you didn't try upstreaming your patches, since you say it would be better to handle things on the Mbed TLS side... I probably misunderstood something, could you clarify?

Here is my patch, 
[mbed_tls_serialize_tls.patch.txt](https://github.com/Mbed-TLS/mbedtls/files/10243699/mbed_tls_serialize_tls.patch.txt). I rebased it yesterday on mbed TLS 3.3.0, it works like a charm, but there are some downsides:

- TLS 1.3 is not supported
- Some logics of buffering is not done inside mbed TLS, and is currently done at Application level. Where for me to properly upstream this patch, these logics should be moved inside mbed TLS stack. Today, the application takes care of buffering the TCP payloads in the order received from the network, in case we don't fully receive the TLS payload in one batch. If it happens, mbed TLS will just inform the application that it wants additional data through `MBEDTLS_ERR_SSL_WANT_READ`. When the application sees this flag, it will serialize the data received alongside the TLS security session, and wait for the next TCP message before calling back mbed TLS with both messages merged into one. Where for me, this buffer should potentially be stored inside the TLS security session.

> Regarding support for serialization before the handshake is complete, it's news to me that this is something people want. I'm a bit surprised, because as I understand it, serialization is mostly useful for long-lived connections to avoid having to re-handshake (as even a resumed handshake is expensive) but it would naively think that (1) the handshake itself would not be spread over a long period of time (like, at most a few minutes for slow devices) and (2) if we didn't complete the handshake yet, then there's less to be saved by resuscitating the SSL context. Apparently my reasoning does not match your experience, so I'm curious to know more.

Today, we can state that during a handshake due to the limitation of not serializing the session before finishing this initial procedure, we can't share the context with other instances of the DTLS/TLS servers. At this stage, we are stateful, meaning that once an instance received a Client Hello, this instance is assigned to the new incoming DTLS/TLS peer to do all the handshake procedure. Whereas once the handshake is finished, we serialized the session, and from this point, any instances of the DTLS/TLS servers can handle post handshake messages from this peer since the context will be shared through a common database.

To summarize, before the end of a handshake procedure, we can't load balance DTLS/TLS messages whereas after the procedure the load balancing is possible across DTLS/TLS servers. To overcome this situation, we have introduced a kind of load balancer to rebalance a handshake message when an instance does not handle any handshake messages from this peer. It works, but it's not a perfect solution since it introduces additional latencies.

> Perhaps you can open two new issues, one for TLS support and one for mid-handshake support, to keep this one focused on version upgrade

I will open it in the coming days, thanks for the suggestion.
